Renesas R-Car: Add PWM driver

pmarzin opened the pull request:
This PR adds the PWM driver to support the PWM controller of R-Car SoC family.

The dts has been kept as much close as Linux as possible.
Unlike the dts, the driver is slightly different as the Zephyr pwm api differs.
Clock divider is updated only when possible.

Actually, the api driver test for the pwm (tests/drivers/pwm/pwm_api/src/test_pwm.c) does not work on R-Car as the test is using prohibited values (0% of duty cycle and cycle value > 1023 are not supported by the R-Car PWM controller).
Nevertheless, it has been tested on H3ULCB with the blinky pwm sample with success.

[UPDATE]
For the pwm_api test, a proposal is to adapt the test in order to not test pwm API with prohibited values. See the related PR here : https://github.com/zephyrproject-rtos/zephyr/pull/48834

Commit:
dts: bindings: pwm: add Renesas R-Car PWM

This add the PWM can bindings for R-Car platforms.
Tested on Renesas Gen3 (h3).

Signed-off-by: Pierre Marzin <pierre.marzin@iot.bzh>

Commit:
drivers: pwm: add Renesas R-Car driver

Add R-Car Gen3 PWM driver.

Clock diviser is automatically adjusted according to requested period
and duty-cycle in order to obtain as much accuracy as possible.
Indeed, in order to improve PWM accurancy, the PWM clock has to fit
the requested period. So use the given period_cycle to define if the
clock as to be adapted. In such case, increase/decrease the clock
diviser to adapt the period_cycle and be sure that it fits into the
10 bits counter of the PWM controller.

Tested on H3ULCB on pwm0 and pwm4.

Signed-off-by: Pierre Marzin <pierre.marzin@iot.bzh>

Commit:
dts: arm: rcar_gen3_cr7: declare pwm0 node

R-Car Gen3 platforms have up to 7 channels. Add the node to
the rcar_gen3_cr7 SoC series. In contrary to Linux, declare
only one PWM controller with 7 channels. So only one node is
written into dtsi file.

Signed-off-by: Pierre Marzin <pierre.marzin@iot.bzh>

Commit:
boards: h3ulcb: Add support of pwm0

Declare pw0 channel0 that can be used on H3ULCB (test pin CP8).
Add support status to H3ULCB board documentation.
PWM0 is disabled by default as its usage could consume uneeded
power.
Only enable it into the test API with overlay.

Signed-off-by: Pierre Marzin <pierre.marzin@iot.bzh>

julien-massot commented on boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7-pinctrl.dtsi:
```diff
@@ -30,4 +30,8 @@
 		pin = <PIN_RX2_A FUNC_RX2_A>;
 	};
+
+	pwm0_pin_out: pwm0_pin_out {
```
```suggestion
	pwm0_default: pwm0_default {
```

pmarzin replied:
Thanks, this has been corrected.

julien-massot commented on drivers/pwm/pwm_rcar.c:
```diff
@@ -0,0 +1,278 @@
+LOG_MODULE_REGISTER(pwm_rcar);
+
+/* Helper macros for PWM */
+#define DEV_PWM_CFG(dev)        ((const struct pwm_rcar_cfg *)(dev)->config)
```
Please drop these helper macros and use dev->data, dev->config directly

pmarzin replied:
Done :)

julien-massot commented on drivers/pwm/pwm_rcar.c:
```diff
@@ -0,0 +1,278 @@
+	uint32_t reg_val;
+	int ret = 0;
+
+	if (pwm != 0) {
```
perhaps add a comment here, that on R-Car we only support one gpio per instance, each of them are declared individually to match Linux dts .

pmarzin replied:
Yes indeed, to be the closest as Linux as possible (regarding the dts), only one pwm channel is allowed per instance. I put a comment. Thanks.

julien-massot commented on drivers/pwm/pwm_rcar.c:
```diff
@@ -0,0 +1,278 @@
+			    uint32_t period_cycles, uint32_t pulse_cycles,
+			    pwm_flags_t flags)
+{
+	const struct pwm_rcar_cfg *config = DEV_PWM_CFG(dev);
```
For this one and all the following
```suggestion
	const struct pwm_rcar_cfg *config = dev->config;
```

julien-massot commented on drivers/pwm/pwm_rcar.c:
```diff
@@ -0,0 +1,278 @@
+				       uint32_t pwm,
+				       uint64_t *cycles)
+{
+	const struct pwm_rcar_cfg *config = DEV_PWM_CFG(dev);
+	struct pwm_rcar_data *data = DEV_PWM_DATA(dev);
```
```suggestion
	const struct pwm_rcar_cfg *config = dev->config;
	struct pwm_rcar_data *data = dev->data;
```

julien-massot commented on dts/arm/renesas/gen3/rcar_gen3_cr7.dtsi:
```diff
@@ -70,6 +71,76 @@
+			label = "pwm3";
+		};
+
+		pwm4: pwm@e6e34000 {
```
"This LSI incorporates seven channels for R-Car H3, R-Car H3-N, R-Car M3-W, R-Car M3-N, R-Car E3; five channels
for R-Car V3M and R-Car V3H; four channels for R-Car D3 pulse width modulation (PWM) timer."

Since all Gen3 SoCs have different number of pwm channels we may move these pwm definition in the SoC specific dts "r8a77951.dtsi".

julien-massot commented on dts/bindings/pwm/renesas,rcar-pwm.yaml:
```diff
@@ -0,0 +1,16 @@
+
+description: Renesas R-Car PWM controller
+
+compatible: "renesas,rcar-pwm"
```
just noticed that Linux is using 
"renesas,pwm-rcar"
https://elixir.bootlin.com/linux/v5.17-rc3/source/Documentation/devicetree/bindings/pwm/renesas,pwm-rcar.yaml
Please renease this file and the compatible entry.

pmarzin replied:
Corrected. Now declared like Linux dts files.

julien-massot commented on dts/arm/renesas/gen3/rcar_gen3_cr7.dtsi:
```diff
@@ -70,6 +71,76 @@
+				<&cpg CPG_CORE CPG_CORE_CLK_S0D12>;
+			status = "disabled";
+			label = "pwm0";
+		};
```
Nits, Linux is describing it in this order:
- compatible
- reg
- clocks
- pwm-cells
- status

julien-massot replied:
https://github.com/torvalds/linux/blob/master/arch/arm64/boot/dts/renesas/r8a77951.dtsi

pmarzin replied:
Corrected, thanks. Now follows the Linux pwm-rcar organization.

julien-massot commented on drivers/pwm/pwm_rcar.c:
```diff
@@ -0,0 +1,278 @@
+
+	/* Disable PWM */
+	pwm_rcar_write_bit(config, RCAR_PWM_CR, RCAR_PWM_CR_EN, false);
+
```
Does by just disabling the pwm you can simulate the 0% duty cycle ?

pmarzin replied:
No, nevertheless, disabling the pwm will set a high level on the channel pin. Like a duty cycle at 100%.

julien-massot commented on drivers/pwm/pwm_rcar.c:
```diff
@@ -0,0 +1,275 @@
+	}
+
+	LOG_DBG("enabled=pwm%d, pulse_cycles=%d, period_cycles=%d,"
+		" duty_cycle=%d", ((config->reg_addr & 0x0000f000) >> 3), pulse_cycles,
```
(config->reg_addr & 0x0000f000) >> 3 ? 

What was the intention ?

pmarzin replied:
According to Table 60.2 from https://www.renesas.com/us/en/document/man/r-car-h3-m3-users-manual documentation, this is to get the PWM channel number, extracted from the reg_addr value.

pmarzin replied:
To simplify, this has been modified. Now, the base address register is printed. Base register will help to identify which PWM channel is used.

mbolivar-nordic reviewed (dismissed):
Devicetree portions LGTM; nice job. I did not review the rest of the driver.

aaillet commented on boards/arm/rcar_h3ulcb/doc/rcar_h3ulcb.rst:
```diff
@@ -166,6 +168,13 @@ I2C is mainly used to manage and power on multiple of onboard chips on the H3ULC
+
+ULCB boards provide one PWM controller with a maximum of 7 channels [0..6]. H3ULCB does provide the pwm0 from test pin CP8 only.
+
+The Kingfisher daughter board is providing the pwm4 channel on the LVDS CN7 connector. Then LVDS backlight could be supported with this PWM driver on Kingfisher through 4th channel.
```
```suggestion
When plugged on a Kingfisher daughter board, pwm4 channel is available on CN7 LVDS connector.
```

pmarzin replied:
Your proposal is easier to understand. Agreed.

aaillet commented on boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml:
```diff
@@ -11,6 +11,7 @@ supported:
   - gpio
   - clock_control
   - uart
+  - pwm
```
If you do not support the pwm test suite (according to your PR message), you should not mark it as supported in your board's yaml file. Otherwise, you will get errors on the pwm suite every time you run twister on your board.
So I think that you should mark this suite as ignored for now before marking it as supported later if you find a way to patch either your driver or the test suite.

pmarzin replied:
Good point here. I will rework this until the test could be passed with success.

aaillet commented on drivers/pwm/pwm_rcar.c:
```diff
@@ -0,0 +1,279 @@
+	/* Prohibiten values */
+	if (period_cycles == 0U || pulse_cycles == 0U ||
+	    pulse_cycles > period_cycles) {
+		return -EINVAL;
```
ENOTSUP as it is not supported by the pwn controller ?

pmarzin replied:
0 is a prohibited value so "Unsupported value" is preferred than "Invalid argument". Correct. Fixed. Thanks :)

aaillet commented on drivers/pwm/pwm_rcar.c:
```diff
@@ -0,0 +1,279 @@
+	 * if period_cycles is less than half of the counter, then the
+	 * clock diviser could be updated as the diviser is a modulo 2.
+	 */
+	if (period_cycles > RCAR_PWM_MAX_CYCLE ||
+	    period_cycles < (RCAR_PWM_MAX_CYCLE / 2)) {
+		LOG_DBG("Adapting frequency diviser...");
+		ret = pwm_rcar_update_clk(config, pwm, &period_cycles, &pulse_cycles);
+		if (ret != 0) {
+			return ret;
+		}
+	}
```
Spaces and tabulation are mixed here, please unify this in the driver file.

pmarzin replied:
Theses spaces are made by uncrustify to align the "if" arguments.

aaillet commented on drivers/pwm/pwm_rcar.c:
```diff
@@ -0,0 +1,279 @@
+		return ret;
+	}
+
+	return 0;
```
`int ret = 0;` then `return ret;` ?

pmarzin replied:
ret is always modified so no need to init it to 0.
I'm ok to simplify the end as `clock_control_get_rate` return 0 if no error.

pmarzin commented:
Hi @mbolivar-nordic ! Some basic improvements and rebase on main have been done (without modification on device tree parts). Can you update your review please ?

julien-massot commented:
@henrikbrixandersen would you like to check this PR ?

henrikbrixandersen commented on drivers/pwm/pwm_rcar.c:
```diff
@@ -0,0 +1,280 @@
+		return -ENOTSUP;
+	}
+
+	/* Prohibiten values */
```
```suggestion
	/* Prohibited values */
```

pmarzin replied:
Fixed! Thanks for your feedback :)

henrikbrixandersen commented on drivers/pwm/pwm_rcar.c:
```diff
@@ -0,0 +1,280 @@
+	/* Prohibiten values */
+	if (period_cycles == 0U || pulse_cycles == 0U ||
+	    pulse_cycles > period_cycles) {
+		return -ENOTSUP;
```
```suggestion
		return -EINVAL;
```

pmarzin replied:
Fixed! Thanks for your feedback :)

mbolivar-nordic reviewed (dismissed):
Looks good to me.

henrikbrixandersen commented on dts/bindings/pwm/renesas,pwm-rcar.yaml:
```diff
@@ -0,0 +1,16 @@
+    "#pwm-cells":
+      const: 2
+
+pwm-cells:
+  - channel
+  - flags
```
```suggestion
pwm-cells:
  - channel
# period in terms of nanoseconds
  - period
  - flags
```

This is the standard convention.

pmarzin replied:
Hi Henrik, thanks. I didn't saw this conventional. I will update it based on the #45013 PR.

pmarzin replied:
Ok, Done ! I did the update with a rebase on main. Do not hesitate to tell me if you see anything else. Thank you @henrikbrixandersen

henrikbrixandersen commented on boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml:
```diff
@@ -14,3 +14,4 @@ supported:
 testing:
   ignore_tags:
     - isotp
+    - pwm
```
I just noticed you are ignoring all `pwm` tests/samples for this board - how come?

pmarzin replied:
As mentioned into the PR message, pwm_api tests use prohibited values that are not hardware supported by our PWM controller. That's why with @aaillet we decided to put it as "ignored" in order to not block twister test suite.
Should we propose a fix into the pwm_api test to add a particular case for this driver ? What is your advise ?
Thanks

henrikbrixandersen replied:
> As mentioned into the PR message, pwm_api tests use prohibited values that are not hardware supported by our PWM controller. That's why with @aaillet we decided to put it as "ignored" in order to not block twister test suite.
> Should we propose a fix into the pwm_api test to add a particular case for this driver ? What is your advise ?

The common PWM tests should be driver agnostic. I'd submit an issue - or even a PR, if you have a proposal for a fix.

pmarzin commented:
The PR is currently blocked by failure tests that are out of PWM scope. They are known and followed by this PR #45102 . Waiting it to solve the tests suite.

pmarzin commented:
As PR #45102 is merged, "update with rebase" has been triggered to solve tests issues. All are good now. This PR is ready :)
@henrikbrixandersen @aaillet  @mbolivar-nordic : sorry for the re-ask : Can you update your approvement please ?

henrikbrixandersen commented on drivers/pwm/pwm_rcar.c:
```diff
@@ -0,0 +1,280 @@
+}
+
+static const struct pwm_driver_api pwm_rcar_driver_api = {
+	.pin_set = pwm_rcar_pin_set,
```
You need to catch up with the recent changes to the PWM API (see 95b56cdffd61f2843b0533d7ab0f1f1ccc9dc092, b317ce8597d471cdc6df117a5cc0877b783f6af7, and related commits)

pmarzin replied:
This PR has been updated with the mentioned commits and some others like https://github.com/zephyrproject-rtos/zephyr/commit/885087b3288346e2e09da92e506e5f390cd4ad29.

Commit:
samples: blinky_pwm: add rcar_h3ulcb_cr7 board overlay's

Use dts overlay to declare the pwm-led only when needed.
This is only used for demo purpose of PWM0 on RCAR gen3
boards (H3ULCB only at the moment).

Signed-off-by: Pierre Marzin <pierre.marzin@iot.bzh>

henrikbrixandersen reviewed (dismissed):
LGTM

aaillet commented on boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml:
```diff
@@ -11,6 +11,7 @@ supported:
   - gpio
   - clock_control
   - uart
+  - pwm
```
Should maybe set to unsupported waiting for #48834 to be merged ? Twister suite will report errors in this state !
What do you think @henrikbrixandersen, it depend on Zephyr test suite politic, maybe you are more aware than me ?

henrikbrixandersen replied:
Either that or exclude the board from the tests for now.

pmarzin replied:
Yes indeed, otherwise the twister test will failed. I removed it !
I did a push force again but everything else remains intact :)

aaillet replied:
Thanks

pmarzin commented:
Sorry, due to a failure test I had to remove "label" property from the dts file.

carlescufi commented:
Merging this without @anangl's approval due to delays in reviews.

pmarzin commented:
🥳 Thanks for your help guys 👍🏼
